### PR TITLE
Implement asynchronous communication mechanism

### PR DESCRIPTION
**Pull request by kibertoad**
This is based on excellent work that @bonw88 did for Review Center, and provides refactored and hardened form of RabbitMQ integration.

ToDo in another PR(s):
* Look into thresholds for requeued message retries, maybe Rabbit already has something?
* Documentation
* Proper implementation of Permissions service (low priority, but let's not have bad examples in the template)

### CODE REVIEW

File: src/modules/users/services/PermissionsService.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Not a real example of a service, do not use for non-testing purposes
+export class PermissionsService {
```
Comment (kibertoad) [inline src/modules/users/services/PermissionsService.ts:2]: this is ugly as hell, but I ran out of time, and this part of the code irrelevant for what this PR is aiming to achieve.

File: .eslintrc.json [hunk 2]
```diff
@@ -62,7 +62,9 @@
     {
       "files": ["*.test.ts", "*.spec.ts"],
       "rules": {
-        "@typescript-eslint/no-non-null-assertion": "off"
+        "@typescript-eslint/no-non-null-assertion": "off",
+        "@typescript-eslint/no-unsafe-return": "off",
+        "@typescript-eslint/no-unsafe-call)": "off"
```
Comment (drdaemos) [inline .eslintrc.json:67]: is `no-unsafe-call)` having extra parenthesis at the end?

Comment (kibertoad) [reply .eslintrc.json]: nice catch, thanks!

File: src/app.ts [hunk 3]
```diff
@@ -219,6 +237,16 @@ export async function getApp(
     } else {
       app.log.info('Skip registering background jobs, test environment')
     }
+
+    if (isAmqpEnabled) {
+      // This is needed to ensure it gets initialized
+      app.diContainer.cradle.amqpConnectionDisposer
```
Comment (elliotykim) [inline src/app.ts:243]: is this a common pattern in our code? is this an artifact of DI?

File: src/infrastructure/amqp/AbstractConsumer.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    })
+
+    await this.channel.assertQueue(this.queueName, {
+      exclusive: false,
+      durable: true,
+      autoDelete: false,
+    })
+  }
+
+  protected deserializeMessage(message: Message | null): Either<'abort', MessagePayloadType> {
```
Comment (kamilwylegala) [inline src/infrastructure/amqp/AbstractConsumer.ts:84]: :large_orange_diamond: Why protected? Can't it's overridden by any child class.

File: src/app.ts (hunk 3 above)
Comment (kibertoad) [reply src/app.ts:243]: It's not, it's related to Awilix creator [refusing to implement eager instantiation](https://github.com/jeffijoe/awilix/issues/310). Probably we'll write some kind of a smarter wrapper over Awilix later to fill in the gaps, because we also miss async instantiators, which Awilix also doesn't want to have.

File: src/infrastructure/amqp/AbstractConsumer.ts (hunk 4 above)
Comment (kibertoad) [reply src/infrastructure/amqp/AbstractConsumer.ts:84]: You are right, this was the case of premature openness. I was assuming that there _might_ be some cases when one class may want to deserialize differently, but then again - YAGNI, and since it's a template, it will be an easy change anyway. 

tl; dr: you are right, will change.

File: src/infrastructure/amqp/AbstractConsumer.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  ) {
+    this.connection = amqpConnection
+    this.errorHandler = consumerErrorResolver
+    this.isShuttingDown = false
+    this.queueName = params.queueName
+    this.messageSchema = params.messageSchema
+    this.newRelicBackgroundTransactionManager = newRelicBackgroundTransactionManager
+  }
+
+  private async init() {
```
Comment (kamilwylegala) [inline src/infrastructure/amqp/AbstractConsumer.ts:52]: :large_orange_diamond: When and where this method should be invoked? I see there is `consume` method but when it is supposed to be called? Is `server` or `app` a good place for that?

Comment (kibertoad) [reply src/infrastructure/amqp/AbstractConsumer.ts:56]: It's a private method, and yes, it's invoked early - `app` does it now, and in consumer tests we also do explicit destroy/init for the cleanup.

File: src/infrastructure/amqp/AbstractConsumer.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+export abstract class AbstractConsumer<MessagePayloadType extends CommonMessage>
+  implements Consumer
+{
+  protected readonly queueName: string
+  protected readonly connection: Connection
+  // @ts-ignore
+  protected channel: Channel
+  protected readonly errorHandler: ConsumerErrorResolver
```
Comment (kamilwylegala) [inline src/infrastructure/amqp/AbstractConsumer.ts:34]: :green_circle: Would be nice to have the same name of the field like the its type: `errorResolver`.

File: src/infrastructure/amqp/ConsumerErrorResolver.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { AmqpMessageInvalidFormat, AmqpValidationError } from './amqpErrors'
+
+export class ConsumerErrorResolver implements ErrorResolver {
+  public processError(error: unknown): InternalError {
+    if (error instanceof Error && error?.name === 'SyntaxError') {
+      return new AmqpMessageInvalidFormat({
+        message: error.message,
+      })
+    }
+    if (error instanceof ZodError) {
```
Comment (kamilwylegala) [inline src/infrastructure/amqp/ConsumerErrorResolver.ts:16]: :large_orange_diamond: Can we expect from `Zod` some kind of versioning mechanism for incoming events?

Or the way it should be: "always keep events backwards compatible".

Comment (kamilwylegala) [reply src/infrastructure/amqp/ConsumerErrorResolver.ts:16]: Thank you Igor, fair enough :blush:

File: src/modules/users/consumers/PermissionConsumer.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    if (!projectUsers || projectUsers.length < message.userIds.length) {
+      // not all users were already created, we need to wait to be able to set permissions
+      return {
+        error: 'retryLater',
+      }
+    }
+
+    // Do not do this in production, some kind of bulk insertion is needed here
+    for (const user of projectUsers) {
+      await this.permissionsService.setPermissions(user.id, message.permissions)
```
Comment (kamilwylegala) [inline src/modules/users/consumers/PermissionConsumer.ts:42]: PHPStorms shows a warning for me here: 
image

Message is: "Unsafe argument of type any assigned to number". Any idea why, it looks that `id` is being resolved to number (by `UserDTO`)

Comment (kibertoad) [reply src/modules/users/consumers/PermissionConsumer.ts:42]: not sure, locally both eslint and webstorm are quiet

File: test/utils/waitUtils.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        .then((result) => {
+          if (result) {
+            resolve(result)
+          } else {
+            retryCount++
+            setTimeout(performCheck, sleepTime)
+          }
+        })
+        .catch((err) => {
+          reject(err)
```
Comment (kamilwylegala) [inline test/utils/waitUtils.ts:25]: Shouldn't we retry on errors too?

File: src/modules/users/consumers/PermissionConsumer.ts (hunk 8 above)
Comment (kamilwylegala) [reply src/modules/users/consumers/PermissionConsumer.ts:42]: ok nvermind, likely it's my setup fault, thanks for confirmation

**Review by kamilwylegala — APPROVED** [review]

File: test/utils/waitUtils.ts (hunk 9 above)
Comment (kibertoad) [reply test/utils/waitUtils.ts:25]: I'm not sure, since this is for using in tests, if we fail with an error somewhere, probably it's fine to fail entire test

**Review by drdaemos — APPROVED** [review]

**Review by lilpoozie2005 — APPROVED** [review]
